codec-dns/src/main/java/io/netty/handler/codec/dns/DefaultDnsRecordDecoder.java

```diff
@@ -0,0 +1,101 @@
+        String name = decodeName(in);
+        DnsRecordType type = DnsRecordType.valueOf(in.readUnsignedShort());
+        DnsRecordClass qClass = DnsRecordClass.valueOf(in.readUnsignedShort());
+        return new DefaultDnsQuestion(name, type, qClass);
+    }
+
+    @Override
+    public <T extends DnsRecord> T decodeRecord(ByteBuf in) throws Exception {
+        final String name = decodeName(in);
+        final DnsRecordType type = DnsRecordType.valueOf(in.readUnsignedShort());
+        final DnsRecordClass aClass = DnsRecordClass.valueOf(in.readUnsignedShort());
+        final long ttl = in.readUnsignedInt();
+        final int length = in.readUnsignedShort();
+        final int offset = in.readerIndex();
+
+        @SuppressWarnings("unchecked")
+        T record = (T) decodeRecord(name, type, aClass, ttl, in, offset, length);
+        in.readerIndex(offset + length);
+        return record;
+    }
+
+    protected DnsRecord decodeRecord(
+            String name, DnsRecordType type, DnsRecordClass dnsClass, long timeToLive,
+            ByteBuf in, int offset, int length) throws Exception {
+
+        return new DefaultDnsRawRecord(
+                name, type, dnsClass, timeToLive, in.duplicate().setIndex(offset, offset + length).retain());
+    }
+
+    /**
+     * Retrieves a domain name given a buffer containing a DNS packet. If the
+     * name contains a pointer, the position of the buffer will be set to
+     * directly after the pointer's index after the name has been read.
+     *
+     * @param in
+     *            the byte buffer containing the DNS packet
+     * @return the domain name for an entry
+     */
+    protected String decodeName(ByteBuf in) {
+        int position = -1;
+        int checked = 0;
+        final int end = in.writerIndex();
+        final StringBuilder name = new StringBuilder(in.readableBytes() << 1);
+        for (int len = in.readUnsignedByte(); in.isReadable() && len != 0; len = in.readUnsignedByte()) {
+            boolean pointer = (len & 0xc0) == 0xc0;
+            if (pointer) {
+                if (position == -1) {
+                    position = in.readerIndex() + 1;
+                }
+
+                final int next = (len & 0x3f) << 8 | in.readUnsignedByte();
+                if (next >= end) {
+                    throw new CorruptedFrameException("name has an out-of-range pointer");
+                }
+                in.readerIndex(next);
+
+                // check for loops
+                checked += 2;
+                if (checked >= end) {
+                    throw new CorruptedFrameException("name contains a loop.");
+                }
+            } else {
+                name.append(in.toString(in.readerIndex(), len, CharsetUtil.UTF_8)).append('.');
+                in.skipBytes(len);
+            }
+        }
+        if (position != -1) {
+            in.readerIndex(position);
+        }
+        if (name.length() == 0) {
+            return "";
```

Comment:
`StringUtil.EMPTY_STRING` does nothing in particular - hard-coded strings are interned, so either way it's just a pointer into the string pool (which, moreover, the compiler will inline).  Unless you expect to need to run Find Usages on "" in an IDE, I don't think there's much point to it.